### PR TITLE
[MI-2474] Added comment modal for approving/rejecting pipeline requests

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**

### CODE REVIEW

File: server/constants/routes.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ const (
 	PathGetGitRepositoryBranches  = "/{organization:[A-Za-z0-9-/ ]+}/{project:[A-Za-z0-9-/ ]+}/repositories/{repository:[A-Za-z0-9-/]+}/branches"
 	PathPipelineReleaseRequest    = "/pipeline_release_request"
 	PathPipelineRunRequest        = "/pipeline_run_request"
+	PathPipelineCommentModal      = "/pipeline_comment_modal"
```
Comment (ayusht2810) [inline server/constants/routes.go:23]: shouldn't the path be `pipeline-comment-modal`?

File: server/constants/constants.go [hunk 2]
```diff
@@ -144,11 +144,16 @@ const (
 
 	PipelineRequestIDApproved          = "approved"
 	PipelineRequestIDRejected          = "rejected"
+	PipelineRequestNameRun             = "run"
```
Comment (Kshitij-Katiyar) [inline server/constants/constants.go:147]: Similar to ```PipelineRequestContextRequestName```
```suggestion
	PipelineRequestRunName             = "run"
```

File: server/constants/constants.go [hunk 3]
```diff
@@ -144,11 +144,16 @@ const (
...
+	PipelineRequestNameRelease         = "release"
```
Comment (Kshitij-Katiyar) [inline server/constants/constants.go:148]:
```suggestion
	PipelineRequestReleaseName         = "release"
```

File: server/plugin/api.go [hunk 4]
```diff
@@ -985,6 +991,66 @@ func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.
 	returnStatusOK(w)
 }
 
+func (p *Plugin) handlePipelineCommentModal(w http.ResponseWriter, r *http.Request) {
+	mattermostUserID := r.Header.Get(constants.HeaderMattermostUserID)
+	decoder := json.NewDecoder(r.Body)
+	postActionIntegrationRequest := &model.PostActionIntegrationRequest{}
+	if err := decoder.Decode(&postActionIntegrationRequest); err != nil {
+		// TODO: prevent posting any error message except oAuth in DM for now and use dialog for all such cases
+		p.handlePipelineApprovalRequestUpdateError("Error decoding PostActionIntegrationRequest params: ", mattermostUserID, err)
```
Comment (ayusht2810) [inline server/plugin/api.go:1000]: shouldn't we use `param` here, as we are only using user ID?

File: server/plugin/api.go [hunk 5]
```diff
@@ -1235,48 +1301,60 @@ func (p *Plugin) handleGetGitRepositoryBranches(w http.ResponseWriter, r *http.R
...
 		p.handlePipelineApprovalRequestUpdateError("Error decoding PostActionIntegrationRequest params: ", mattermostUserID, err)
 		p.handleError(w, r, &serializers.Error{Code: http.StatusInternalServerError, Message: err.Error()})
 		return
 	}
 
-	requestType := postActionIntegrationRequest.Context[constants.PipelineRequestContextRequestType].(string)
+	values := strings.Split(submitRequest.State, "$")
+	organization := values[0]
+	projectName := values[1]
+	requestType := values[3]
```
Comment (ayusht2810) [inline server/plugin/api.go:1315]: don't we need to check the length here?

File: server/plugin/api.go [hunk 6]
```diff
@@ -1304,28 +1382,34 @@ func (p *Plugin) handlePipelineApproveOrRejectRunRequest(w http.ResponseWriter,
...
 
-	organization := postActionIntegrationRequest.Context[constants.PipelineRequestContextOrganization].(string)
-	projectID := postActionIntegrationRequest.Context[constants.PipelineRequestContextProjectID].(string)
-	approvalID := postActionIntegrationRequest.Context[constants.PipelineRequestContextApprovalID].(string)
-	requestType := postActionIntegrationRequest.Context[constants.PipelineRequestContextRequestType].(string)
+	values := strings.Split(submitRequest.State, "$")
+	organization := values[0]
+	projectID := values[1]
+	approvalID := values[2]
+	requestType := values[3]
```
Comment (ayusht2810) [inline server/plugin/api.go:1394]: same

File: server/plugin/api.go [hunk 7]
```diff
@@ -985,6 +991,66 @@ func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.
 	returnStatusOK(w)
 }
 
+func (p *Plugin) handlePipelineCommentModal(w http.ResponseWriter, r *http.Request) {
+	mattermostUserID := r.Header.Get(constants.HeaderMattermostUserID)
+	decoder := json.NewDecoder(r.Body)
+	postActionIntegrationRequest := &model.PostActionIntegrationRequest{}
+	if err := decoder.Decode(&postActionIntegrationRequest); err != nil {
+		// TODO: prevent posting any error message except oAuth in DM for now and use dialog for all such cases
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:999]:
```suggestion
		// TODO: prevent posting any error messages except oAuth in DM for now and use dialog for all such cases
```

File: server/constants/constants.go (hunk 2 above)
Comment (raghavaggarwal2308) [reply server/constants/constants.go:158]: I think what we have is better. As `run` is the value of field `RequestName`

File: server/plugin/client.go [hunk 8]
```diff
@@ -299,11 +300,7 @@ func (c *client) UpdatePipelineRunApprovalRequest(pipelineApproveRequestPayload
 
 	var pipelineRunApproveResponse *serializers.PipelineRunApproveResponse
 	_, statusCode, err := c.CallJSON(c.plugin.getConfiguration().AzureDevopsAPIBaseURL, pipelineApproveRunRequestURL, http.MethodPatch, mattermostUserID, &pipelineApproveRequestPayload, &pipelineRunApproveResponse, nil)
-	if err != nil {
-		return nil, statusCode, err
-	}
-
-	return pipelineRunApproveResponse, statusCode, nil
+	return pipelineRunApproveResponse, statusCode, err
```
Comment (ayusht2810) [inline server/plugin/client.go:303]: why this change?

File: server/plugin/client.go [hunk 9]
```diff
@@ -520,6 +517,11 @@ func (c *client) Call(basePath, method, path, contentType string, mattermostUser
 	return responseData, resp.StatusCode, fmt.Errorf("errorMessage %s", errResp.Message)
 }
 
+func (c *client) OpenDialogRequest(body *model.OpenDialogRequest, mattermostUserID string) (int, error) {
+	_, statusCode, err := c.CallJSON(c.plugin.getConfiguration().MattermostSiteURL, "/api/v4/actions/dialogs/open", http.MethodPost, mattermostUserID, body, nil, nil)
```
Comment (ayusht2810) [inline server/plugin/client.go:521]: create a constant for the path

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api.go [hunk 10]
```diff
@@ -1187,6 +1193,66 @@ func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.
...
 }
 
+func (p *Plugin) handlePipelineCommentModal(w http.ResponseWriter, r *http.Request) {
+	mattermostUserID := r.Header.Get(constants.HeaderMattermostUserID)
+	decoder := json.NewDecoder(r.Body)
+	postActionIntegrationRequest := &model.PostActionIntegrationRequest{}
+	if err := decoder.Decode(&postActionIntegrationRequest); err != nil {
+		// TODO: prevent posting any error messages except oAuth in DM for now and use dialog for all such cases
+		p.handlePipelineApprovalRequestUpdateError("Error decoding PostActionIntegrationRequest param: ", mattermostUserID, err)
+		p.handleError(w, r, &serializers.Error{Code: http.StatusInternalServerError, Message: err.Error()})
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:1203]: this should internal server error or a bad request (400 series) as we are not able to decode ?

File: server/plugin/api.go [hunk 11]
```diff
@@ -1187,6 +1193,66 @@ func (p *Plugin) handleSubscriptionNotifications(w http.ResponseWriter, r *http.
...
+
+	organization := postActionIntegrationRequest.Context[constants.PipelineRequestContextOrganization].(string)
+	requestType := postActionIntegrationRequest.Context[constants.PipelineRequestContextRequestType].(string)
+
+	elements := []model.DialogElement{
+		{
+			DisplayName: "Comment:",
+			Name:        constants.DialogFieldNameComment,
+			Type:        "text",
+			Optional:    true,
```
Comment (Kshitij-Katiyar) [inline server/plugin/api.go:1226]: whats the default value for optional and check if it is true remove this line

Comment (raghavaggarwal2308) [reply server/plugin/api.go:1226]: default value is false

File: server/plugin/api.go (hunk 10 above)
Comment (raghavaggarwal2308) [reply server/plugin/api.go:1203]: This API is called by us internally. I guess that's why we are using `InternalServerError` here.
Can you please confirm @avas27JTG

**Review by Kshitij-Katiyar — APPROVED** [review]
